arch/riscv/kernel/devicetree.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Enables use of Device Tree for RISC-V peripheral devices.
+ *
+ * Author: Jamey Hicks <jamey.hicks@gmail.com>
+ */
+
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/slab.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <linux/irqchip.h>
+#include <linux/of_fdt.h>
+#include <linux/of_platform.h>
+
+int __init devicetree_init(void)
+{
+	resource_size_t initial_dtb = 0x04100000;
```

Comment:
Agreed.

Hardcoding the address was definitely a quick hack. I should at least have
used a CONFIG option. My intention is to connect this in a more appropriate
way.

I do not know why I used obj-y and not obj-$(CONFIG_OF) in the Makefile.

I think the boot loader should be able to pass in the address of the device
tree, either on the command line or via a register. What do you recommend?

-Jamey


On Fri, Jan 22, 2016 at 2:31 AM Albert Ou <notifications@github.com> wrote:

> In arch/riscv/kernel/devicetree.c
> <https://github.com/riscv/riscv-linux/pull/30#discussion_r50508237>:
>
> > + */
> > +
> > +#include <linux/module.h>
> > +#include <linux/types.h>
> > +#include <linux/kernel.h>
> > +#include <linux/init.h>
> > +#include <linux/slab.h>
> > +#include <linux/device.h>
> > +#include <linux/io.h>
> > +#include <linux/irqchip.h>
> > +#include <linux/of_fdt.h>
> > +#include <linux/of_platform.h>
> > +
> > +int __init devicetree_init(void)
> > +{
> > +	resource_size_t initial_dtb = 0x04100000;
>
> We should avoid hard-coding a physical address for the FDT as well as
> directly assigning to initial_boot_params. It appears that the usual
> approach is to pass a pointer to early_init_dt_verify(), which might
> either be provided by the bootloader or statically initialized to
> __dtb_start when using the builtin FDT.
>
> More generally, device tree handling doesn't seem appropriate for a
> standalone kernel module. It should probably be part of setup_arch(), as
> in most other architectures, preferably written in a way that respects
> CONFIG_OF.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/riscv/riscv-linux/pull/30/files#r50508237>.
>